scripts/circleci_scraper/scraper.py

```diff
@@ -0,0 +1,348 @@
+
+        Raises:
+            CircleCIClientError: If there is an error in the CircleCI API request.
+            CircleCIScraperError: If there is an error in downloading the artifacts.
+        """
+        for pipeline_config in pipeline_configs:
+            self.logger.info(f"Scrape {pipeline_config.organization}/{pipeline_config.repository}")
+            self.export_test_metadata_and_artifacts_by_pipeline(pipeline_config, date_limit)
+
+    def export_test_metadata_and_artifacts_by_pipeline(
+        self, pipeline_config: CircleCIScraperPipelineConfig, date_limit: datetime | None = None
+    ) -> None:
+        """Export test metadata and artifacts for a single pipeline.
+
+        Args:
+            pipeline_config (CircleCIScraperPipelineConfig): A pipeline configuration.
+            date_limit (datetime | None): The date limit for fetching data. Defaults to None.
+
+        Raises:
+            CircleCIClientError: If there is an error in the CircleCI API request.
+            CircleCIScraperError: If there is an error in downloading the artifacts.
+        """
+        next_page_token: str | None = None
+        while True:
+            pipelines: PipelineGroup = self._client.get_pipelines(
+                pipeline_config.organization, next_page_token
+            )
+            for pipeline in pipelines.items:
+                # Filter for branches, usually we only observe 'main' or equivalent
+                if (
+                    pipeline_config.branches
+                    and pipeline.vcs
+                    and pipeline.vcs.branch
+                    and pipeline.vcs.branch not in pipeline_config.branches
+                ):
+                    continue
+                # Filter for date limit
+                created_at_datetime = datetime.strptime(
+                    pipeline.created_at, "%Y-%m-%dT%H:%M:%S.%fZ"
+                ).replace(tzinfo=timezone.utc)
+                if date_limit and created_at_datetime < date_limit:
+                    return
+                self.export_test_metadata_and_artifacts_by_pipeline_id(
+                    pipeline.id,
+                    pipeline_config.organization,
+                    pipeline_config.repository,
+                    pipeline_config.workflows,
+                )
+            next_page_token = pipelines.next_page_token
+            if not next_page_token:
+                break
+
+    def export_test_metadata_and_artifacts_by_pipeline_id(
+        self,
+        pipeline_id: str,
+        organization: str,
+        repository: str,
+        workflow_configs: dict[str, list[str]],
+    ) -> None:
+        """Export test metadata and artifacts for a specific pipeline ID.
+
+        Args:
+            pipeline_id (str): The pipeline ID.
+            organization (str): The organization name.
+            repository (str): The repository name.
+            workflow_configs (WorkflowConfigs): The workflow configurations.
```

Comment:
Similarly, L141 says `dict[str, list[str]]` but here we say `WorkflowConfigs` (which looks to be the only Ctrl+F result for "WorkflowConfigs", unless i have some files hidden by default.